crates/hotshot/src/traits/networking/push_cdn_network/definition/message_hook.rs

```diff
@@ -0,0 +1,482 @@
+            if global_bps != 0 && local_bps > max_allowed_bps {
+                error!(
+                    "Local bps ({}) is greater than maximum allowed bps ({}), kicking user",
+                    local_bps, max_allowed_bps
+                );
+                return Err(anyhow::anyhow!(
+                    "Local bps ({}) is greater than maximum allowed bps ({}), kicking user",
+                    local_bps,
+                    max_allowed_bps
+                ));
+            }
+
+            // Reset the check time
+            sample.last_checked_time = Instant::now();
+        }
+
+        // Commit the sample if that interval has elapsed
+        if sample.last_committed_time.elapsed() >= self.sample_commit_interval {
+            sma.commit_sample(sample);
+        }
+
+        Ok(())
+    }
+
+    /// Process incoming broadcast messages from the user
+    fn process_broadcast_message(&mut self, broadcast: &mut Broadcast) -> Result<HookResult> {
+        // Process through the `SMA`
+        self.process_against_sma(broadcast.message.len(), MessageType::Broadcast)?;
+
+        // Calculate the hash of the message
+        let mut hasher = Hash64::default();
+        hasher.write(&broadcast.message);
+        hasher.write(&broadcast.topics);
+
+        // Make sure we have not already seen it
+        if self.message_hash_cache.put(hasher.finish(), ()).is_some() {
+            return Ok(HookResult::SkipMessage);
+        }
+
+        // Make sure it is deserializable
+        // let (_, _) = Self::deserialize_message(&broadcast.message)?;
+
+        Ok(HookResult::ProcessMessage)
+    }
+
+    /// Process incoming direct messages from the user
+    fn process_direct_message(&mut self, direct: &mut Direct) -> Result<HookResult> {
+        // Process through the `SMA`
+        self.process_against_sma(direct.message.len(), MessageType::Direct)?;
+
+        // Calculate the hash of the message
+        let mut hasher = Hash64::default();
+        hasher.write(&direct.message);
+        hasher.write(&direct.recipient);
+
+        // Make sure we have not already seen it
+        if self.message_hash_cache.put(hasher.finish(), ()).is_some() {
+            return Ok(HookResult::SkipMessage);
+        }
+
+        // Make sure it is deserializable
+        // let (_, _) = Self::deserialize_message(&direct.message)?;
+
+        Ok(HookResult::ProcessMessage)
+    }
+
+    // fn deserialize_message(message: &[u8]) -> Result<(Message<T>, Version)> {
+    //     // Hack off the version
+    //     let (version, message) =
+    //         Version::deserialize(&message).with_context(|| "failed to deserialize message")?;
+
+    //     // Deserialize the message
+    //     let message = Serializer::<StaticVersion<0, 1>>::deserialize_no_version(&message)
+    //         .with_context(|| "failed to deserialize message")?;
+
+    //     // Return the version and message
+    //     Ok((message, version))
+    // }
+}
```

Comment:
i see some stuff commented out will this be used in the future or can we clean up?

Reply:
Yeah - we may want to use it in the future.

Right now the hooks are completely agnostic to HotShot because of serialization/versioning. If we want to know what kind of HotShot message the user sent, we have to basically run consensus to figure out what version we should be asking for

In the future this may become more important (e.g. only let a node vote once per view or something) but we're not doing it quite yet